Adding download policy testing permutations.

bherrin3 opened the pull request:
Cases for no policy, change in policy and invalid policy.

Closes#4220
https://pulp.plan.io/issues/4220

dralley committed:
Add content filter

re #3784
https://pulp.plan.io/issues/3784

Commit:
Updated for pulp and pulpcore-plugin changes

ref #4192
https://pulp.plan.io/issues/4192

Commit:
Rename cache_only to streamed

https://pulp.plan.io/issues/4243
re #4243

dkliban committed:
Merge pull request #138 from daviddavis/issue4192

Updating to use pulpcore-plugin repo

Commit:
Fixing duplicate content exception

fixes #3436
https://pulp.plan.io/issues/3436

bherrin3 committed:
Adding download policy testing permutations.

Cases for no policy, change in policy and invalid policy.

Closes#4220
https://pulp.plan.io/issues/4220

bherrin3 commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +127,52 @@ def test_all(self):
+
+    def test_03_invalid_policy(self):
+        """Verify an invalid policy does not update the remote policy."""
+        self.remote.update(self.client.get(self.remote['_href']))
```
```suggestion
        
```
Should be removed. Already tested.

nixocio commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -16,7 +16,6 @@
     FILE_REMOTE_PATH
 )
 from pulp_file.tests.functional.utils import gen_file_remote, skip_if
-from pulp_file.tests.functional.utils import set_up_module as setUpModule  # noqa:F401
```
Why are we removing this?

bherrin3 replied:
The linter picked up that setUpModule was not used. I didn't see a use case for it within the file. 

Is there a usage of this that I am missing? 

All tests passed with the removal of the line.

```
(pulp3) [herring@redherring api]$ pylint test_crud_remotes.py 
************* Module pulp_file.tests.functional.api.test_crud_remotes
test_crud_remotes.py:19:0: W0611: Unused set_up_module imported from pulp_file.tests.functional.utils as setUpModule (unused-import)

-----------------------------------
Your code has been rated at 9.90/10

(pulp3) [herring@redherring api]$
```

bherrin3 replied:
Had a discussion about this. Understood this is now part of the module setup.

Re-adding

nixocio commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +127,52 @@ def test_all(self):
+
+    This test targets the following issues:
+
+    * `Pulp #4420 <https://pulp.plan.io/issues/4420>`_
```
What about add the steps described in the issue itself?

nixocio commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +127,52 @@ def test_all(self):
+
+    This test targets the following issues:
+
+    * `Pulp #4420 <https://pulp.plan.io/issues/4420>`_
```
```suggestion
This test does the following:

1. Describe action...
2. Describe assertion...

    * `Pulp #4420 <https://pulp.plan.io/issues/4420>`_
```

bherrin3 replied:
Agree. Added more comments to what is being tested and where more information can be found.

nixocio commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +127,52 @@ def test_all(self):
+        """Verify default policy when no policy is defined."""
+        del self.body['policy']
+        self.remote.update(self.client.post(FILE_REMOTE_PATH, self.body))
+        self.assertEqual(self.remote['policy'], 'immediate')
```
What about add a third parameter `msg` in case of test failure?

bherrin3 replied:
Yup. That was a miss. Thanks!

nixocio commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +127,52 @@ def test_all(self):
+        self.body['policy'] = choice(remove_policy)
+        self.client.patch(self.remote['_href'], self.body)
+        self.remote.update(self.client.get(self.remote['_href']))
+        self.assertEqual(self.remote['policy'], self.body['policy'])
```
What about add a third parameter `msg` in case of test failure?

bherrin3 replied:
Adding to all assertEqual evals. Thanks.

bherrin3 commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +128,82 @@ def test_all(self):
+    def tearDownClass(cls):
+        """Clean class-wide variable."""
+        cls.client.delete(cls.remote['_href'])
+
```
@kersommoura as we discussed, should the test methods AFTER `test_01_no_defined_policy` be decorated with the `@skip_if(bool, 'remote', False)` in this Class to avoid dependent test failure?

See below, if I understand correctly.

bherrin3 commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +128,82 @@ def test_all(self):
+        del self.body['policy']
+        self.remote.update(self.client.post(FILE_REMOTE_PATH, self.body))
+        self.assertEqual(self.remote['policy'], 'immediate', self.remote)
+
```
Basing the fact if there is a failure, `remote` would be cleaned, and there is no reason to run further tests.

```suggestion
@skip_if(bool, 'remote', False)
```

bherrin3 commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +128,82 @@ def test_all(self):
+        self.client.patch(self.remote['_href'], {'policy': changed_policy})
+        self.remote.update(self.client.get(self.remote['_href']))
+        self.assertEqual(self.remote['policy'], changed_policy, self.remote)
+
```
Again, skip if remote is note defined.

```suggestion
@skip_if(bool, 'remote', False)
```

nixocio committed:
Simplify test CRUD remotes

Remove unnecessary repo creation as part of the CRUD remotes tests.

'[noissue]'

bmbouter committed:
Merge pull request #140 from bmbouter/rename-policy

Rename cache_only to streamed

bherrin3 commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +128,82 @@ def test_all(self):
+        """
+        remove_policy = DOWNLOAD_POLICIES.copy()
+        remove_policy.remove('immediate')
+        changed_policy = choice(remove_policy)
```
@kersommoura talked with bruno, and he had the suggestion of removing L186-188 with:
```suggestion
        self.client.patch(self.remote['_href'], {'policy': choice([item for item in DOWNLOAD_POLICIES if item != 'immediate'])})
```

As long as in `constants.py` I changed the `DOWNLOAD_POLICIES` from a `list` to a `tuple`.

The only other code to use `DOWNLOAD_POLICIES` in this project was L223. Therefore, the change should not be an issue.

Let me know if you have any comments.

nixocio replied:
That is a nice improvement. This constant DOWNLOAD_POLICIES was added to Pulp-Smash, as a list. See: https://github.com/PulpQE/pulp-smash/pull/1156

This will avoid different plugins have to define the same constant.

We need the `changed_policy` in order to make the final assertion, then:
```suggestion
changed_policy = choice(
      [item for item in DOWNLOAD_POLICIES if item != 'immediate']
)
self.client.patch(self.remote['_href'], {'policy': changed_policy })
```

bherrin3 replied:
hehe... that is JUST what I got done implementing and formatting (reading PEP-8 [0]).

Get on vacation, @kersommoura ! This isn't critical! ;)

Thank you!

[0] - https://www.python.org/dev/peps/pep-0008/#multiline-if-statements

bherrin3 replied:
@kersommoura , do you then suggest we change the import of `DOWNLOAD_POLICIES` in `test_crud_remotes.py` to `pulp_smash` and removal of this line from `constants.py`? 

If so, I can open up a PR after this to address that change. 

Still, not sure why even there it was added as a list, but that `horse is out of the barn`. 

Thanks for your time, and enjoy your vacation!

nixocio commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +128,82 @@ def test_all(self):
+
+    This test targets the following issues:
+
+    * `Pulp #4420 <https://pulp.plan.io/issues/4420>`_
```
Very nice docstring. Very well explained.

nixocio commented on pulp_file/tests/functional/api/test_crud_remotes.py:
```diff
@@ -128,6 +128,82 @@ def test_all(self):
+        with self.assertRaises(HTTPError):
+            self.client.patch(self.remote['_href'], {'policy': utils.uuid4()})
+        self.remote.update(self.client.get(self.remote['_href']))
+        self.assertEqual(remote['policy'], self.remote['policy'], self.remote)
```
Very clear as well.
